designs/SpanMarshallers.md

```diff
@@ -0,0 +1,262 @@
+
+Alternatively, the `MarshalUsingAttribute` could provide a `Type ElementNativeType { get; set; }` property instead of an `ElementIndrectionLevel` property and support specifying the native type of the element of the collection this way. However, this design would block support for marshalling collections of collections.
+
+#### Example: Using generic collection marshalling for spans
+
+This design could be used to provide a default marshaller for spans and arrays. Below is an example simple marshaller for `Span<T>`. This design does not include all possible optimizations, such as stack allocation, for simpilicity of the example.
+
+```csharp
+[GenericCollectionMarshaller]
+public ref struct SpanMarshaler<T>
+{
+    private Span<T> managedCollection;
+
+    private int nativeElementSize;
+
+    public SpanMarshaler(Span<T> collection, int nativeSizeOfElements)
+    {
+       managedCollection = collection;
+       Value = Marshal.AllocCoTaskMem(collection.Length * nativeSizeOfElements);
+       nativeElementSize = nativeSizeOfElements;
+    }
+
+    public Span<T> ManagedValues => managedCollection;
+
+    public void SetUnmarshalledCollectionLength(int length)
+    {
+       managedCollection = new T[value];
+    }
+
+    public IntPtr Value { get; set; }
+
+    public unsafe Span<byte> NativeValueStorage => MemoryMarshal.CreateSpan(ref *(byte*)(Value), Length);
+
+    public Span<T> ToManaged() => managedCollection;
+
+    public void FreeNative()
+    {
+      if (Value != IntPtr.Zero)
+      {
+          Marshal.FreeCoTaskMem(Value);
+      }
+    }
+}
+```
+
+This design could also be applied to support the built-in array marshalling if it is desired to move that marshalling out of the stub and into shared code.
+
+#### Future extension to the above model: Non-contiguous collection support
+
+If a managed or native representation of a collection has a non-contiguous element layout, then developers currently will need to convert to or from array/span types at the interop boundary. This section proposes an API that would enable developers to convert directly between a managed and native non-contiguous collection layout as part of marshalling.
+
+An `IsContiguousCollection` boolean property could be added to the `GenericCollectionMarshaller` attribute, with a default value of `true`. When the value is `false`, that would specify that the collection is noncontiguous in either managed or native representations. Then additional methods should be added to the generic collection model, and some methods would be removed:
+
+```diff
+- [GenericCollectionMarshaller]
++ [GenericCollectionMarshaller(IsContiguousCollection = false)]
```

Comment:
How about renaming `GenericCollectionMarshaller` to be `GenericContiguousCollectionMarshaller`? We can then call the non-contiguous one something else.